v0/destinations/hs/transform.js

```diff
@@ -1,244 +1,609 @@
-    } else {
-      endpoint = "https://api.hubapi.com/contacts/v1/contact";
-    }
-    response.method = defaultPostRequestConfig.requestMethod;
-    response.body.JSON = removeUndefinedValues(payload);
+  let endpoint;
+  if (contactId) {
+    // update
+    endpoint = IDENTIFY_CRM_UPDATE_NEW_CONTACT.replace(":contactId", contactId);
   } else {
-    params = removeUndefinedValues(payload);
+    // create
+    endpoint = IDENTIFY_CRM_CREATE_NEW_CONTACT;
   }
+
+  const response = defaultRequestConfig();
+  response.endpoint = endpoint;
+  response.method = defaultPostRequestConfig.requestMethod;
   response.headers = {
     "Content-Type": "application/json"
   };
-  response.endpoint = endpoint;
-  response.userId = message.anonymousId;
-  response.params = params;
-  response.statusCode = 200;
+  response.body.JSON = removeUndefinedAndNullValues(payload);
+
+  // choosing API Type
+  if (Config.authorizationType === "newPrivateAppApi") {
+    // Private Apps
+    response.headers = {
+      ...response.headers,
+      Authorization: `Bearer ${Config.accessToken}`
+    };
+  } else {
+    // use legacy API Key
+    response.params = { hapikey: Config.apiKey };
+  }
 
   return response;
-}
+};
 
-async function processTrack(message, destination, propertyMap) {
-  const parameters = {
-    _a: destination.Config.hubID,
-    _n: message.event
-  };
+/**
+ * using legacy API
+ * Reference:
+ * https://legacydocs.hubspot.com/docs/methods/contacts/create_contact
+ * https://legacydocs.hubspot.com/docs/methods/contacts/create_or_update
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} propertyMap
+ * @returns
+ */
+const processLegacyIdentify = async (message, destination, propertyMap) => {
+  const { Config } = destination;
+  const traits = getFieldValueFromMessage(message, "traits");
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  // if mappedToDestination is set true, then add externalId to traits
+  if (mappedToDestination) {
+    addExternalIdToTraits(message);
+  }
 
-  if (
-    message.properties &&
-    (message.properties.revenue || message.properties.value)
-  ) {
-    // eslint-disable-next-line dot-notation
-    parameters["_m"] = message.properties.revenue || message.properties.value;
+  if (!traits || !traits.email) {
+    throw new CustomError(
+      "[HS]:: Identify without email is not supported.",
+      400
+    );
   }
+
   const userProperties = await getTransformedJSON(
     message,
-    hSIdentifyConfigJson,
+    hsCommonConfigJson,
     destination,
     propertyMap
   );
 
-  return responseBuilderSimple(
-    { ...parameters, ...userProperties },
+  const payload = {
+    properties: formatPropertyValueForIdentify(userProperties)
+  };
+
+  const { email } = traits;
+  let endpoint;
+  if (email) {
+    endpoint = IDENTIFY_CREATE_UPDATE_CONTACT.replace(":contact_email", email);
+  } else {
+    endpoint = IDENTIFY_CREATE_NEW_CONTACT;
+  }
```

Comment:
We will be using batch endpoints for identify, is there any specific need that you are thinking we will need normal endpoints to be assigned?

Reply:
normal endpoints are used when router transform is not being used.

Reply:
yes, but hubspot is marked as router transform in config be and included in `features.json` as well. So AFAIK events by default will land up in router transform. Can there be any other scenario which I am missing? @ItsSudip @utsabc

Reply:
They can change their config variable from the server if the user has limited resources.